pkg/bundle/internal/truststore/types.go

```diff
@@ -98,10 +99,14 @@ func (e pkcs12Encoder) Encode(trustBundle *util.CertPool) ([]byte, error) {
 		})
 	}
 
-	encoder := pkcs12.LegacyRC2
+	encoder := pkcs12.Passwordless
 
-	if e.password == "" {
-		encoder = pkcs12.Passwordless
+	if e.password != "" {
+		encoder = pkcs12.LegacyRC2.
+			// Short-circuiting the rand generator to make our PKCS#12 truststores deterministic.
+			// This should allow use of unconditional SSA requests from controller.
+			// See: https://cert-manager.io/docs/faq/#why-are-passwords-on-jks-or-pkcs12-files-not-helpful
+			WithRand(rand.New(rand.NewSource(1))) //#nosec G404
```

Comment:
Can this behaviour be made optional/tunable by the user/administrator? I imagine some environments may require use of a secure randomness source even if the underlying algorithm is considered insecure, though I'm not sure. What does other software in this space do?

Reply:
> Can this behaviour be made optional/tunable by the user/administrator?

No, not without losing the opportunity to greatly simplify the controller (drop needsUpdate-logic). Dropping a secure random source here is for sure controversial. But I think it could be worth it to simplify our code. Some additional arguments:

- Our password-less PKCS#12 (the default) is deterministic.
- We already stated (in our FAQ) that we support keystores/truststores for legacy reasons, and not because we consider them the recommended approach.
- We already got [a suggestion](https://github.com/cert-manager/trust-manager/issues/428) to include the truststore password just next to the truststore in the target configmap/secret. If we decide to implement anything in this direction, a requirement for a secure random source becomes even more meaningless.

> What does other software in this space do?

No idea. I am pretty sure that any software producing PKCS#12 truststores with a password attempts to use a secure random source. But a password-less PKCS#12 is the new standard for truststores in Java - after they abandoned JKS.